### PR TITLE
This PR adds the custom search screen

### PR DESCRIPTION
**Pull request by Lumiere-Lodi**
This PR adds the custom search screen and it is using dynamic field feature for the available fields you can add. 

**NB:** 

We first must merge #27 before we merge this one. 
There are changes that has been implemented in #27 already. To avoid repetition we did not feature them here.

### CODE REVIEW

**Review by IlyesBenAmara — CHANGES_REQUESTED** [review]
LGTM, i left you a single comment

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: src/components/customSearch/CustomSearch.tsx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+const CustomSearch: React.FC = () => {
+  const { availableFields } = useAppConfig()
+  console.log(availableFields)
+  //TODO: find a better way of handling error while posting the search request
+  const postSearchQuery = useMutation({
+    mutationFn: ApiClient.postCustomSearchQuery,
+    onError: (error: AxiosError) => {
+      console.log(`Oops! Error getting search result: ${error.message}`)
+    }
+  })
```
[the remark this answers is not in the document]
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearch.tsx:25]: We are not using error directly inside the useMutation() function. We are not using it outside. So I have only get mutate from the function. Thanks.

File: src/components/customSearch/CustomSearch.tsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+import { useAppConfig } from '../../hooks/useAppConfig'
+import ApiClient from '../../services/ApiClient'
+import { CustomSearchQuery, FlagLabel } from '../../types/SimpleSearch'
+import SearchFlags from '../search/SearchFlags'
+import PageHeader from '../shell/PageHeader'
+import CustomSearchRow from './CustomSearchRow'
+
+const CustomSearch: React.FC = () => {
+  const { availableFields } = useAppConfig()
+  console.log(availableFields)
```
[the remark this answers is not in the document]
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearch.tsx:18]: Thanks

**Review by IlyesBenAmara — APPROVED** [review]
LGTM :+1:

File: src/components/customSearch/CustomSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                                startIcon={<AddIcon />}
+                                onClick={() => {
+                                  push({
+                                    parameters: [initialCustomSearchValues]
+                                  })
+                                }}
+                                sx={{ fontSize: '13px' }}
+                              >
+                                Add Group
+                              </Button>
```
Comment (marrouchi) [inline src/components/customSearch/CustomSearch.tsx:182]:
```suggestion
                              <Button
                                variant="text"
                                startIcon={<AddIcon />}
                                onClick={() => {
                                  push({
                                    parameters: [initialCustomSearchValues]
                                  })
                                }}
                                sx={{ fontSize: '13px' }}
                              >
                                Add Group
                              </Button>
```
Could we refactor this code into a seperate component that we could use for "Add field" and "Add group" ?

File: src/components/customSearch/CustomSearchRow.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            options={availableFields}
+            title={'Select Type'}
+            description={'Select Field Type'}
+            sx={{ width: 220 }}
+            name={`$or[${fieldGroupIndex}].parameters[${index}].fieldName`}
+
+          />
+        </Grid>
+        <Grid item>
+          {fieldName === 'dob' ? (
```
Comment (marrouchi) [inline src/components/customSearch/CustomSearchRow.tsx:93]:
```suggestion
          {fieldType === 'Date' ? (
```

File: src/components/customSearch/CustomSearchRow.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            <SearchDateInput
+              label={fieldLabel || 'Select a field type'}
```
Comment (marrouchi) [inline src/components/customSearch/CustomSearchRow.tsx:95]:
```suggestion
              label={fieldLabel}
```
The fieldLabel should always be available ;)

File: src/components/customSearch/CustomSearchRow.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  index: number
+  onChange?: (e: React.ChangeEvent<any>) => void
+  remove: <T>(index: number) => T | undefined
+  enableDelete?: boolean
+  enableCondition?: boolean
+  fieldGroupIndex: number
+  setFieldValue: (field: string, value: any, shouldValidate?: boolean | undefined) => void
+}
+
+const CustomSearchRow: React.FC<CustomSearchRowProps> = ({
```
Comment (marrouchi) [inline src/components/customSearch/CustomSearchRow.tsx:22]: Could we re-use the search row we used for the simple search ?

**Review by marrouchi — CHANGES_REQUESTED** [review]
@Lumiere-Lodi Looks good, here some initial comments :)

File: src/components/customSearch/CustomSearchRow.tsx (hunk 4 above)
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearchRow.tsx:93]: @marrouchi This will not work for our use case. We are using a state the change the field type once the user has selected from the select dropdown. To trigger the onChange function on Select your value has to change and update the FieldName state then change the second field type accordingly. If we use fieldType the problem is some fields have the same type. So it will not trigger the onChange function in the Select component. That is the reason why we are using fieldName. Let me know what you think about it or has a better solution.

File: src/components/customSearch/CustomSearch.tsx (hunk 3 above)
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearch.tsx:182]: Thanks

File: src/components/customSearch/CustomSearchRow.tsx (hunk 6 above)
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearchRow.tsx]: done

File: src/components/customSearch/CustomSearch.tsx [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+                            direction="column"
+                            width="100%"
+                            alignItems={'center'}
+                            sx={{ mt: 1 }}
+                          >
+                            <Grid
+                              item
+                              container
+                              direction="row"
+                              width="756px"
```
Comment (IlyesBenAmara) [inline src/components/customSearch/CustomSearch.tsx:161]: Can we use responsive values instead of a fixed width ? (maxWidth, minWidth; 'sm, 'md', 'lg')

**Review by IlyesBenAmara — CHANGES_REQUESTED** [review]
Good job overall, left a couple of comments.

File: src/components/customSearch/CustomSearch.tsx [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+                breadcrumbs={[
+                  {
+                    icon: <MoreHorizOutlined />
+                  },
+                  {
+                    icon: <SearchIcon />,
+                    title: 'Search'
+                  }
+                ]}
+              />
```
[the remark this answers is not in the document]
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearch.tsx:62]: Hi @IlyesBenAmara 

I just looked at the LinkedRecords and the PatientDetails screen as well where we have implemented this

We have two different use case here. The way you are using in LinkedRecords or PatientDetails caters for a single or multiple buttons but not for toggleButton. The reason why We have them separate here is we are using toggleButton and not multiple individual buttons. So the toggleButton component used here should be correct for this use case.

File: src/components/customSearch/CustomSearch.tsx [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+                    height: '42px',
+                    width: '172px',
+                    borderColor: theme => theme.palette.primary.main
+                  }}
+                  href={'/search'}
+                >
+                  <Typography variant="button">SIMPLE SEARCH</Typography>
+                </Button>
+              </Grid>
+            </Grid>
```
[the remark this answers is not in the document]
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearch.tsx:94]: I left a reply above. Let me know what you think. 

Thanks

File: package.json [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "jempi-web",
   "version": "0.1.0",
+  "type": "module",
```
Comment (marrouchi) [inline package.json:4]: @Lumiere-Lodi Why did you need to add this ?

File: src/App.tsx [hunk 11]
```diff
@@ -56,6 +57,7 @@ const routes: Route[] = [
         element: <MatchDetails />
       },
       { path: 'search', element: <SimpleSearch /> },
+      {path: '/custom-search', element: <CustomSearch/>},
```
Comment (marrouchi) [inline src/App.tsx:60]: Let's lint this ;)

File: src/components/customSearch/AddFieldOrGroupButton.tsx [hunk 12]
```diff
@@ -0,0 +1,31 @@
+import { Button } from '@mui/material'
+import { SearchParameter } from '../../types/SimpleSearch'
+import AddIcon from '@mui/icons-material/Add'
+
+interface AddFieldOrGroupButtonProps {
+  push: (obj: any) => void
```
Comment (marrouchi) [inline src/components/customSearch/AddFieldOrGroupButton.tsx:6]:
```suggestion
  onClick: (obj: any) => void
```
What does the push func do ? Can we give it a better name ? Also why do we need to resend the prop ?

File: src/components/customSearch/AddFieldOrGroupButton.tsx [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  initialCustomSearchValues: SearchParameter
+  label: string
+}
+const AddFieldOrGroupButton: React.FC<any> = ({
+  push,
+  initialCustomSearchValues,
+  label
+}) => {
+  return (
+    <>
```
Comment (marrouchi) [inline src/components/customSearch/AddFieldOrGroupButton.tsx:16]: We don't need fragment unless there's multiple elements.

File: src/components/customSearch/CustomSearch.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (marrouchi) [reply src/components/customSearch/CustomSearch.tsx:62]: I think what Ilyes means is that there's a "buttons" prop in the PageHeader where you could set the search flags.

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: src/components/customSearch/CustomSearchRow.tsx (hunk 4 above)
Comment (marrouchi) [reply src/components/customSearch/CustomSearchRow.tsx]: The "dob" is a dynamic field, it could not exist for a specific implementation. So you cannot use it ...

File: package.json (hunk 10 above)
Comment (Lumiere-Lodi) [reply package.json:4]: my bad. I was testing something with Matt. No worry

File: src/components/customSearch/AddFieldOrGroupButton.tsx (hunk 12 above)
Comment (Lumiere-Lodi) [reply src/components/customSearch/AddFieldOrGroupButton.tsx:6]: The push function is coming from Formik. That's the function that adds a new index to the array.

File: src/components/customSearch/AddFieldOrGroupButton.tsx (hunk 13 above)
Comment (Lumiere-Lodi) [reply src/components/customSearch/AddFieldOrGroupButton.tsx:16]: sure

File: src/components/customSearch/AddFieldOrGroupButton.tsx [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  onClick,
+  initialCustomSearchValues,
+  label
+}) => {
+  return (
+      <Button
+        variant="text"
+        startIcon={<AddIcon />}
+        onClick={() => {
+          onClick(initialCustomSearchValues)
```
Comment (marrouchi) [inline src/components/customSearch/AddFieldOrGroupButton.tsx:20]: I think we don't need to pass "initialCustomSearchValues" as a prop and then sent it back on click ?
Should rather be `onClick={onClick}` and then handle that in the parent component.

File: src/components/customSearch/CustomSearch.tsx [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+const CustomSearch: React.FC = () => {
+  //TODO: find a better way of handling error while posting the search request
+  const { mutate } = useMutation({
+    mutationFn: ApiClient.postCustomSearchQuery,
+    onError: (error: AxiosError) => {
+      console.log(`Oops! Error getting search result: ${error.message}`)
+    }
+  })
+
+  const initialCustomSearchValues: SearchParameter = {
```
Comment (marrouchi) [inline src/components/customSearch/CustomSearch.tsx:25]:
```suggestion
  const initialSearchParameter: SearchParameter = {
```

File: src/components/search/SearchRow.tsx [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+  ]
+
+  const [newField, setNewField] = useState<DisplayField | undefined>(undefined)
+  const [matchType, setMatchType] = useState<string>('')
+
+  const handleFieldNameChange = (event: SelectChangeEvent) => {
+
+    const field = availableFields.find(obj => obj.fieldName === event.target.value)
+
+    console.log(field)
```
Comment (marrouchi) [inline src/components/search/SearchRow.tsx]: Let remove any left console.log

File: src/components/search/SearchRow.tsx [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+              onChange={handleFieldNameChange}
+              index={index}
+              options={availableFields.map(value => ({
+                value: value.fieldName,
+                label: value.fieldLabel
+              }))}
+              title={'Select Type'}
+              description={'Select Field Type'}
+              sx={{ width: 220, mt: 0.4 }}
+              name={`$or[${fieldGroupIndex}].parameters[${index}].fieldName`}
```
Comment (marrouchi) [inline src/components/search/SearchRow.tsx:98]: I think we need to pass the name of the field as a prop and remove all the others : fieldGroupIndex, index, isCustomRow ...

File: src/types/SimpleSearch.ts [hunk 18]
```diff
@@ -4,13 +4,18 @@ export interface SearchParameter {
...
 
-export interface CustomSearchParameters extends SearchParameter {
-  operator: string
-}
-
 export interface CustomSearchQuery {
-  parameters: CustomSearchParameters[]
+  parameters: SearchParameter[]
+}
+export interface $or {
```
Comment (marrouchi) [inline src/types/SimpleSearch.ts:10]:
```suggestion
export interface CustomSearchQuery {
```

File: src/types/SimpleSearch.ts [hunk 19]
```diff
@@ -4,13 +4,18 @@ export interface SearchParameter {
   distance: number
 }
 
-export interface CustomSearchParameters extends SearchParameter {
-  operator: string
-}
-
 export interface CustomSearchQuery {
```
Comment (marrouchi) [inline src/types/SimpleSearch.ts:7]:
```suggestion
export interface SimpleSearchQuery {
```

**Review by marrouchi — CHANGES_REQUESTED** [review]
Looks better now, left a couple of comments ;)

File: src/components/customSearch/CustomSearch.tsx (hunk 15 above)
Comment (Lumiere-Lodi) [reply src/components/customSearch/CustomSearch.tsx:25]: sure

File: src/types/SimpleSearch.ts (hunk 18 above)
Comment (Lumiere-Lodi) [reply src/types/SimpleSearch.ts:10]: done

File: src/components/search/SearchRow.tsx (hunk 17 above)
Comment (Lumiere-Lodi) [reply src/components/search/SearchRow.tsx:98]: sure

**Review by marrouchi — APPROVED** [review]

**Review by IlyesBenAmara — APPROVED** [review]
LGTM :+1:
